### PR TITLE
feat(compiler-cli): automatically lower lambda expressions in metadata

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

When a lambda expression is used in an Angular decorator an error is reported during AOT requesting the user rewrite the expression.

## What is the new behavior?

The expression is automatically rewritten to an exported reference.

NOTE: This is not yet enabled in `ngc` as it requires the transform based compiler which is not yet the default.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  return visitSourceFile(sourceFile);
+}
+
+export function getExpressionLoweringTransformFactory(requestsMap: RequestsMap):
+    (context: ts.TransformationContext) => (sourceFile: ts.SourceFile) => ts.SourceFile {
+  // Return the factory
+  return (context: ts.TransformationContext) => (sourceFile: ts.SourceFile): ts.SourceFile => {
+    const requests = requestsMap.getRequests(sourceFile);
+    if (requests && requests.size) {
+      const min = Array.from(requests.keys()).reduce((p, c) => c < p ? c : p);
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: consider using `Math.min` and ensure values are num

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  private ensureMetadataAndRequests(sourceFile: ts.SourceFile): MetadataAndLoweringRequests {
+    let result = this.metadataCache.get(sourceFile.fileName);
+    if (!result) {
+      result = this.getMetadataAndRequests(sourceFile);
+      this.metadataCache.set(sourceFile.fileName, result);
+    }
+    return result;
+  }
+
+  private getMetadataAndRequests(sourceFile: ts.SourceFile, ): MetadataAndLoweringRequests {
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: trailing comma

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    const requests = new Map<number, LoweringRequest>();
+    const replaceNode =
+        (node: ts.Node) => {
+          const name = freshIdent();
+          requests.set(node.pos, {name, kind: node.kind, location: node.pos, end: node.end});
+          return {__symbolic: 'reference', name};
+        }
+
+    const substituteExpression = (value: MetadataValue, node: ts.Node): MetadataValue => {
+      if (node.kind === ts.SyntaxKind.ArrowFunction) {
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: handle function expression

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+      if (declarations.length) {
+        inserts.push({priorTo: result, declarations})
+      }
+      return result;
+    }
+
+    const traversedSource = ts.visitEachChild(sourceFile, topLevelStatement, context);
+    if (inserts.length) {
+      // Insert the declarations before the statement that was rewritten reference them.
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: typo: references

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  return visitSourceFile(sourceFile);
+}
+
+export function getExpressionLoweringTransformFactory(requestsMap: RequestsMap):
+    (context: ts.TransformationContext) => (sourceFile: ts.SourceFile) => ts.SourceFile {
+  // Return the factory
+  return (context: ts.TransformationContext) => (sourceFile: ts.SourceFile): ts.SourceFile => {
+    const requests = requestsMap.getRequests(sourceFile);
+    if (requests && requests.size) {
+      // Calculate the range of intersting locations. The transform will only visit nodes in this
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: Should with move that to the transform itself ? 
_(seems like this leaks an impl detail to me)_

File: packages/compiler-cli/test/ngc_spec.ts [hunk 6]
```diff
@@ -308,14 +308,130 @@ describe('ngc command-line', () => {
...
+        const mymoduleSource = fs.readFileSync(mymodulejs, 'utf8');
+        expect(mymoduleSource).toContain('var ɵ0 = function () { return new Foo(); }');
+        expect(mymoduleSource).toContain('export { ɵ0');
+
+        const mymodulefactory = path.resolve(outDir, 'mymodule.ngfactory.js');
+        const mymodulefactorySource = fs.readFileSync(mymodulefactory, 'utf8');
+        expect(mymodulefactorySource).toContain('"someToken", i1.ɵ0');
+      });
+
+      it('should able to multiple expressions', () => {
```
Comment (vicb) [inline packages/compiler-cli/test/ngc_spec.ts]: typo: `should able to multiple expressions`

File: packages/compiler-cli/test/ngc_spec.ts [hunk 7]
```diff
@@ -308,14 +308,130 @@ describe('ngc command-line', () => {
...
+              {provide: 'someToken', useFactory: () => new Foo()},
+              {provide: 'someToken', useFactory: () => new Foo()},
+              {provide: 'someToken', useFactory: () => new Foo()}
+            ]
+          })
+          export class MyModule {}
+        `);
+        expect(compile()).toEqual(0);
+        const mymodulejs = path.resolve(outDir, 'mymodule.js');
+        const mymoduleSource = fs.readFileSync(mymodulejs, 'utf8');
```
Comment (vicb) [inline packages/compiler-cli/test/ngc_spec.ts]: nit: `, 'utf8')`  is the default IIRC

File: packages/compiler-cli/src/transformers/lower_expressions.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    if (inserts.length) {
+      // Insert the declarations before the statement that was rewritten reference them.
+      const insertMap = toMap(inserts, i => i.priorTo);
+      const newStatements: ts.Statement[] = [...traversedSource.statements];
+      for (let i = newStatements.length; i >= 0; i--) {
+        const statement = newStatements[i];
+        const insert = insertMap.get(statement);
+        if (insert) {
+          const declarations = insert.declarations.map(
+              i => ts.createVariableDeclaration(
```
Comment (vicb) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: createConst ?

File: packages/compiler-cli/test/ngc_spec.ts (hunk 7 above)
Comment (chuckjaz) [inline packages/compiler-cli/test/ngc_spec.ts]: Actually it returns a `Buffer` if you don't specify a character set.

File: packages/compiler-cli/src/transformers/lower_expressions.ts (hunk 8 above)
Comment (chuckjaz) [inline packages/compiler-cli/src/transformers/lower_expressions.ts]: Change this to be const by adding a const modifier to the declaration list.

File: packages/compiler-cli/test/ngc_spec.ts (hunk 7 above)
Comment (vicb) [inline packages/compiler-cli/test/ngc_spec.ts]: oh right

File: packages/compiler-cli/test/transformers/lower_expressions_spec.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+      return ts.forEachChild(node, find);
+    }
+  }
+  return ts.forEachChild(node, find);
+}
+
+function normalizeResult(result: string): string {
+  // Remove TypeScript prefixes
+  // Remove new lines
+  // Squish adjacent spaces
```
Comment (vicb) [inline packages/compiler-cli/test/transformers/lower_expressions_spec.ts]: typo ?

**Review by vicb — APPROVED** [review]

File: packages/compiler-cli/test/transformers/lower_expressions_spec.ts (hunk 9 above)
Comment (chuckjaz) [inline packages/compiler-cli/test/transformers/lower_expressions_spec.ts]: Squish and squash are pretty much synonyms.

**Pull request merged**